modules/gapi/include/opencv2/gapi/gtransform.hpp

```diff
@@ -0,0 +1,202 @@
+struct TransType
+{
+    using type = T;
+};
+
+template <typename T>
+using is_nongapi_type2 = std::is_same<T, typename TransType<T>::type>;
+
+template <typename T>
+typename std::enable_if<!is_nongapi_type2<T>::value,
+                        typename TransType<T>::type>::type
+get_in_args(const GArgs &in_args, int idx)
+{
+    return util::get<typename TransType<T>::type>(in_args.at(idx));
+}
+
+template <typename T>
+typename std::enable_if<is_nongapi_type2<T>::value, T>::type get_in_args(const GArgs &in_args, int idx)
+{
+    return in_args.at(idx).template get<T>();
+}
+
+//////////////////////////////////////////////////////////////////////
+
+struct GAPI_EXPORTS GTransform
+{
+    using F = std::function<GArgs(const GArgs &)>;
+
+    const char *description;
+    F pattern;
+    F substitute;
+
+    explicit GTransform(const char *d, const F &p, const F &s) : description(d), pattern(p), substitute(s){};
+};
+
+template <class Impl>
+GTransform transformation()
+{
+    return GTransform(Impl::descr(), &Impl::get_pattern, &Impl::get_substitute);
+}
+
+////////////////////////////////////////////////////////////////////
+
+template <typename, typename, typename>
+struct TransHelper;
+
+// FIXME: code duplication:
+// consider better approach like in compound kernels with tuple wrappers and context class
+
+template <typename K, typename... Ins, typename Out>
+struct TransHelper<K, std::tuple<Ins...>, Out>
+{
+    // FIXME: code duplication
+    template <int... IIs>
+    static GArgs get_pattern_impl(const GArgs &in_args, detail::Seq<IIs...>)
+    {
+        using R = typename TransType<Out>::type;
+        const R r = K::pattern(get_in_args<Ins>(in_args, IIs)...);
+        return GArgs{GArg(r)};
+    }
+
+    template <int... IIs>
+    static GArgs get_substitute_impl(const GArgs &in_args, detail::Seq<IIs...>)
+    {
+        using R = typename TransType<Out>::type;
+        const R r = K::substitute(get_in_args<Ins>(in_args, IIs)...);
+        return GArgs{GArg(r)};
+    }
+
+    static GArgs get_pattern(const GArgs &in_args)
+    {
+        return get_pattern_impl(in_args, typename detail::MkSeq<sizeof...(Ins)>::type());
+    }
+    static GArgs get_substitute(const GArgs &in_args)
+    {
+        return get_substitute_impl(in_args, typename detail::MkSeq<sizeof...(Ins)>::type());
+    }
+};
+
+template <typename K, typename... Ins, typename... Outs>
+struct TransHelper<K, std::tuple<Ins...>, std::tuple<Outs...>>
+{
+    // FIXME: code duplication
+    template <int... IIs, int... OIs>
+    static GArgs get_pattern_impl(const GArgs &in_args, detail::Seq<IIs...>, detail::Seq<OIs...>)
+    {
+        using R = std::tuple<typename TransType<Outs>::type...>;
+        const R r = K::pattern(get_in_args<Ins>(in_args, IIs)...);
+        return GArgs{GArg(std::get<OIs>(r))...};
+    }
+
+    template <int... IIs, int... OIs>
+    static GArgs get_substitute_impl(const GArgs &in_args, detail::Seq<IIs...>, detail::Seq<OIs...>)
```

Comment:
Pattern and substitute have the same signature. Need to figure out how to get rid of copy-paste.

Reply:
something like this should work (and if we always operate with `GArg`/`GArgs` this may be a good candidate for universal utility):

```cpp
template<typename F, int... IIs, int... OIs>
static GArgs get_impl(const GArgs &in_args, detail::Seq<IIs...>, detail::Seq<OIs...>) {
        using R = std::tuple<Outs...>;
        const R r = F(in_args.at(IIs).template get<Ins>()...);
        return GArgs{GArg(std::get<OIs>(r))...};
}
```
usage:
```cpp
    static GArgs get_pattern(const GArgs &in_args)
    {
        return get_impl(K::pattern, in_args, typename detail::MkSeq<sizeof...(Ins)>::type(),
                                typename detail::MkSeq<sizeof...(Outs)>::type());
    }
    static GArgs get_substitute(const GArgs &in_args)
    {
        return get_impl(K::substitute, in_args, typename detail::MkSeq<sizeof...(Ins)>::type(),
                                   typename detail::MkSeq<sizeof...(Outs)>::type());
    }
```

Also, the whole thing looks like [invoke](https://en.cppreference.com/w/cpp/utility/functional/invoke) (since C++17, though), might be better to find good standard alternative

Reply:
remark:
```cpp
static GArgs get_impl(F f, const GArgs &in_args, detail::Seq<IIs...>, detail::Seq<OIs...>) {
  const R r = f(in_args.at(IIs).template get<Ins>()...);
...
}
```

Reply:
Done, thanks!